framework/scripts/generate_test_list.py

```diff
@@ -74,32 +79,79 @@ def main():
-        if default_name is not None:
-            f.write(f'    if (!test_list_name) return {default_name}_test_list;\n')
+            f.write(f'static constexpr struct test * const {name}_test_list_[] = {{')
+            test_list = [ f'\n    &_test_{test}' for test in tests ]
+            f.write(','.join(test_list))
+            f.write('\n};\n')
+            path = test_list_files[name]
+            f.write(f'static const struct TestList {name}_test_list = {{ {"cpu_" + name if name != "default" else 0}, "{name}", "{path}", {name}_test_list_ }};\n\n')
+
+        # create default "null" list if not defined in the option
+        if default_name is None:
+            f.write(f'static const struct TestList default_test_list = {{ 0, "default", nullptr, {{}} }};\n\n')
+
+
+        # function to check if the list is selected, applied to non-default lists only
+        if other_names:
+            f.write('static bool test_list_matches(const TestList* test_list, const char* name)\n')
+            f.write('{\n')
+            f.write('#ifdef SANDSTONE\n')
+            f.write('    if ((name == nullptr) || (strcmp(name, "auto") == 0)) {\n')
+            f.write('        return (test_list->features & cpu_features) == test_list->features;\n')
+            f.write('    }\n')
+            f.write('#endif\n')
+            f.write('    return (name != nullptr) && (strcmp(name, test_list->name) == 0);\n')
+            f.write('}\n\n')
+
+        # selector function
+        f.write('const TestList& select_test_list(const char *name)\n')
+        f.write('{\n')
         for name in test_lists.keys():
-            f.write(f'    if  (!strcmp(test_list_name, "{name}")) return {name}_test_list;\n')
-        f.write('    return nullopt;\n')
+            if default_name is None or name != default_name:
+                f.write(f'    if (test_list_matches(&{name}_test_list, name)) return {name}_test_list;\n')
+
+        # emit warning if neither list matches
+        if default_name is not None:
+            f.write('    if (name && (strcmp(name, "default") != 0)) {\n')
+            f.write('        logging_printf(LOG_LEVEL_QUIET, "# ERROR: list for %s not found. Using default\\n", name);\n')
+            f.write('    }\n')
+        f.write('    return default_test_list;\n')
         f.write('}\n')
 
     # generate .h file
     with open(h_file, 'w') as f:
+        f.write("// **This is generated file, do not edit**\n\n")
+        f.write('#include "sandstone.h"\n')
         f.write('#include <optional>\n')
-        f.write('#include <span>\n')
-        f.write('#include "sandstone.h"\n\n')
-        f.write('std::optional<const std::span<struct test * const>> get_test_list(const char *);\n')
+        f.write('#include <span>\n\n')
+
+        # per-gen list definition
+        f.write('typedef struct TestList {\n')
+        f.write('    uint64_t features;\n')
+        f.write('    const char* name;\n')
+        f.write('    const char* filename;\n')
```

Comment:
What's the point of the file name? It won't exist in the user's system.

Reply:
I'd assume, that it might be used to identify exact list being used in execution (I don't expect "trivial" mapping where e.g. clx => data/clx.txt, rather expect data/clx_ppv_240226.txt). I don't know details of your "requirement" on metadata assigned to each list.
In the code it is used to determine, whether the list is NULL (the list is not created on any file) or empty. (o/c it might be "fixed" with optional in the structure).

Reply:
BTW, should we consider optimization, where the test pointers are kept per file (instead of current "per name")?

Reply:
Replaced "test list per generation" with "shared test lists" to optimize memory usage with duplicated files. See snippets.

Reply:
Ah, "which test list" I had meant "clx", not the file name.